src/config.rs

```diff
@@ -0,0 +1,313 @@
+            Err(ConfigError::YamlParse(meta.clone()))
+        }
+    }
+}
+
+macro_rules! yaml_array {
+    ($n:expr => $($i:expr)+) => {
+        impl<T: FromYaml> FromYaml for [T; $n] {
+            fn from_yaml(meta: &ConfigMeta, config: &Yaml) -> Result<Self, ConfigError> {
+                if let &Yaml::Array(ref array) = config {
+                    if array.len() != $n {
+                        return Err(ConfigError::YamlParse(meta.clone()));
+                    }
+
+                    Ok([
+                       $(
+                            try!(T::from_yaml(meta, &array.get($i).unwrap())
+                                .map_err(|_| ConfigError::YamlParse(meta.clone()))),
+                       )+
+                    ])
+                }
+                else {
+                    Err(ConfigError::YamlParse(meta.clone()))
+                }
+            }
+        }
+    }
+}
+
+yaml_array!(1 => 0);
+yaml_array!(2 => 0 1);
+yaml_array!(3 => 0 1 2);
+yaml_array!(4 => 0 1 2 3);
+yaml_array!(5 => 0 1 2 3 4);
+yaml_array!(6 => 0 1 2 3 4 5);
+yaml_array!(7 => 0 1 2 3 4 5 6);
+yaml_array!(8 => 0 1 2 3 4 5 6 7);
+yaml_array!(9 => 0 1 2 3 4 5 6 7 8);
+yaml_array!(10 => 0 1 2 3 4 5 6 7 8 9);
+
+pub trait FromFile: Sized {
+    // From a file relative to current config
+    fn from_file_raw(meta: &ConfigMeta, path: &Path) -> Result<Self, ConfigError>;
+
+    // From a file relative to project
+    fn from_file(path: &Path) -> Result<Self, ConfigError> {
+        Self::from_file_raw(&ConfigMeta::default(), path)
+    }
+}
+
+impl<T: FromYaml + Sized> FromFile for T {
+    fn from_file_raw(meta: &ConfigMeta, path: &Path) -> Result<T, ConfigError> {
+        let mut next_meta = meta.clone();
+        let mut field_path = meta.path.parent().unwrap_or(Path::new("")).to_path_buf();
+
+        field_path.push(path);
+
+        if field_path.is_dir() && field_path.exists() {
+            field_path.push("config");
+        }
+
+        field_path.set_extension("yml");
+        next_meta.path = field_path;
+
+        if next_meta.path.exists() {
+            let mut file = try!(File::open(next_meta.path.as_path())
+                .map_err(|e| ConfigError::FileError(next_meta.path.as_path().display().to_string(), e)));
+            let mut buffer = String::new();
+
+            try!(file.read_to_string(&mut buffer)
+                .map_err(|e| ConfigError::FileError(next_meta.path.as_path().display().to_string(), e)));
+
+            let yaml = try!(YamlLoader::load_from_str(&buffer)
+                .map_err(|e| ConfigError::YamlScan(e)));
+            let hash = &yaml[0];
+
+            <T>::from_yaml(&next_meta, hash)
+        }
+        else {
+            Err(ConfigError::MissingExternalFile(next_meta.clone()))
+        }
+    }
+}
+
+macro_rules! config {
+    ($root:ident {
+        $( $field:ident: $ty:ty = $name:expr, )*
+    }) => {
+
+        #[derive(Clone, Debug)]
+        pub struct $root {
+            _meta: ConfigMeta,

```

Comment:
Should I move this to be completely invisible/in the background to the user or keep it exposed? It may be useful if we want to do something like writing to the configs.